File: SDK/CSharp/Device.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;

namespace SensorCloud
{
	public class Device
	{
		private readonly string _deviceId;
		private IRequests _requests;

		/// <summary>
		/// Initializes a new instance of a SensorCloud Device for a specific account.
		/// </summary>
		/// <param name="deviceId">id for the Device's SensorCloud account</param>
		/// <param name="key">key required to access the Device</param>
		/// <param name="authServer">The authentication server to use.
		/// <remarks>default: https://sensorcloud.microstrain.com</remarks>
		/// </param>
		public Device(String deviceId, String key, String authServer = "https://sensorcloud.microstrain.com", IRequests requests = null)
		{
			//if a request was passed in use it.  If one wasn't passed in then we create a default RequestsFatory
			_requests = requests != null? requests : new SensorCloudRequests(deviceId, key, authServer);
			
			this._deviceId = deviceId;
		}

		/// <summary>
		/// Determine if a sensor exists for this Device.
		/// </summary>
		/// <param name="sensorName"></param>
		/// <returns>True if the sensor exists.</returns>
		public bool HasSensor(String sensorName)
		{
			var request = _requests.url("/sensors/" + sensorName + "/")
					.Param("version", "1")
					.Accept("application/xdr")
					.Get();
			
			// check the response code for success
			switch(request.ResponseCode)
			{
				case HttpStatusCode.OK: return true;
				case HttpStatusCode.NotFound: return false;
				default: throw SensorCloudException.GenerateSensorCloudException(request, "HasSensor failed");
			}
		}

	
		/// <summary>
		/// Add a new sensor to the Device.
		/// </summary>
		/// <param name="sensorName">Name that will be given to the sensor.  Once a sensor is created it's name cannot be changed.
		/// <remarks>sensorName must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]</remarks>
		/// </param>
		public Sensor AddSensor(String sensorName)
		{
			return AddSensor(sensorName, "", "", "");
		}

		/// <summary>
		/// Add a new sensor to the Device.
		/// </summary>
		/// <param name="sensorName">Name that will be given to the sensor.  Once a sensor is created it's name cannot be changed.
		/// <remarks>sensorName must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]</remarks>
		/// </param>
		/// <param name="label">User Friendly label for the sensor.  May be changed later.
		/// <remarks>label supports unicode, and must not be more than 50 bytes when encoded as UTF-8.</remarks>
		/// </param>
		/// <param name="type">Type of the sensor. May be changed later.
		/// <remarks>type supports unicode, and must not be more than 50 bytes when encoded as UTF-8.</remarks>
		/// </param>
		/// <param name="description">Description of the sensor.  May be changed later.
		/// <remarks>description supports unicode, and must not be more than 1000 bytes when encoded as UTF-8.</remarks>
		/// </param>
		public Sensor AddSensor(String sensorName, String label, String type, String description)
		{
			var payload = new MemoryStream();
			var xdr = new XdrWriter(payload);
			
			int VERSION = 1;
			xdr.WriteInt(VERSION);
			xdr.WriteString(type);
			xdr.WriteString(label);
			xdr.WriteString(description);
			
			var request = _requests.url("/sensors/" + sensorName + "/")
								   .Param("version", "1")
								   .ContentType("application/xdr")
								   .Data(payload.ToArray())
								   .Put();
			
			// check the response code for success
			if (request.ResponseCode != HttpStatusCode.Created)
			{
				throw SensorCloudException.GenerateSensorCloudException(request, "AddSensor Failed");
			}

			return new Sensor(sensorName, _requests);
		}

		/// <summary>
		/// Get an existing Sensor.  GetSensor will succeed even if the sensor doesn't exit.
		/// However,  you will get a SensorDoesntExist exception when you call a function on the returned
		/// Sensor object that communicates  with SensorCloud.
		/// </summary>
		/// <param name="sensorName"></param>
		/// <returns></returns>
		public Sensor GetSensor(string sensorName)
		{
			return new Sensor(sensorName, _requests);
		}

		/// <summary>
		/// Get a list of existing Sensors for this device.
		/// </summary>
		public List<Sensor> GetSensors()
		{
			List<Sensor> outputList = new List<Sensor>();
			//make a get request for channel attributes to determine if the channel exists
			var request = _requests.url("/sensors/")
														.Param("version", "1")
														.Accept("application/xdr")
														.Get();

			// check the response code for success
			if (request.ResponseCode == HttpStatusCode.OK)
			{
				var ms = new MemoryStream(request.Raw);
				var xdr = new XdrReader(ms);
				//The first value is the version.  We expect this to be 1
				int version = xdr.ReadInt();
				if (version != 1)
				{
					throw new SensorCloudException("Unsupported xdr version");
				}

				//The next value is an integer telling us how many sensors to unpack
				//Each sensor has its info packed a certain way, so this allows us to know how many sensors we must unpack
				int sensorCount = xdr.ReadInt();
				//loop through sensors
				for(var i=0; i < sensorCount; i++)
				{
					string sensorName = xdr.ReadString(1000);
					//for example's sake, I'm going to push a Sensor Object onto our output array
					//This object only needs the Sensor name, and will fetch the Sensor info on it's own
					//However, to get all sensor names we must still parse all of the stream.  You can modify 
					//this function to meet your own needs, and return a different set of information 
					//tailored to what you need.
					outputList.Add(new Sensor(sensorName, _requests));

					string sensortype = xdr.ReadString(1000);
					string sensorlabel = xdr.ReadString(1000);  //a token is generally about 60 chars.  Limit the read to at most 1000 chars as a precation so in their is a protocol error we don't try to allocate lots of memory
					string sensordescripiton = xdr.ReadString(1000);

					//the next value is the number of channels this sensor has
					int channelCount = xdr.ReadInt();
					//loop all channels
					for (var j=0; j < channelCount; j++)
					{
						string channelName = xdr.ReadString(1000);
						string channelLabel = xdr.ReadString(1000);
						string channelDescription = xdr.ReadString(1000);

						//This tells us how many datastreams the channel has
						//This should be one, as we only currently support TimeSeries, but could grow as more stream types are added
						int streamCount = xdr.ReadInt();
						//loop through streams
						for (var k=0; k < streamCount; k++)
						{
							//the first value in the stream is the type of stream.  We expect this to be TS_V1, meaning timeseries version 1
							string streamType = xdr.ReadString(1000);
							//the second value of a stream is the number of bytes that describes the stream.  
							//This is convenient as it allows us to just skip the stream if we are not interested in it.  
							//For this example's purpose, we are uninterested and will simply skip ahead in the xdr.  We do this by reading the number bytes, but doing anything
							//with the data
							int totalBytes = xdr.ReadInt();
							//just reading the bytes by previously parsed length advances the xdr reader forward in the data
							xdr.ReadBytes(totalBytes);
						}
					}
				}
				return outputList;
			}
			else
			{
				throw SensorCloudException.GenerateSensorCloudException(request, "Get Sensors Failed");
			}
		}


		/// <summary>
		/// Delete's a Sensor from the device.  All channels must be deleted before you can delete the Sensor.
		/// </summary>
		/// <param name="sensorName"></param>
		public void DeleteSensor(string sensorName)
		{
			var request = _requests.url("/sensors/" + sensorName + "/")
								   .Param("version", "1")
								   .Delete();

			if(request.ResponseCode != HttpStatusCode.NoContent)
			{
				throw SensorCloudException.GenerateSensorCloudException(request, "Delete Sensor Failed");
			}
		}
	}
}

File: Example Code/Java/com/sensorcloud/Device.java
/*
 * Copyright 2012 MicroStrain Inc. All Rights Reserved.
 *
 * Distributed under the Simplified BSD License.
 * See file license.txt
 *
 */

package com.sensorcloud;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.List;

import com.sensorcloud.Requests.Request;


public class Device
{
	private String authServer;
	private String deviceId;
	private String key;
	
	private String apiServer;
	private String authToken;
	
	/**
	 * @param deviceId  this is the unique ID for a particular device on SensorCloud
	 * @param key  the key generated for API access to this device.    
	 */
	public Device(String deviceId, String key)
	{
		final String DEFAULT_AUTH_SERVER = "https://sensorcloud.microstrain.com";
		
		this.authServer = DEFAULT_AUTH_SERVER;
		this.deviceId = deviceId;
		this.key = key;
	}
	
	/**
	 * @param deviceId  this is the unique ID for a particular device on SensorCloud
	 * @param key  the key generated for API access to this device.    
	 * @param authServer - custom auth-server to authenticate with.
	 */
	public Device(String deviceId, String key, String authServer)
	{
		this.authServer = authServer;
		this.deviceId = deviceId;
		this.key = key;
	}
		
	
	/**
	 * Authenticates using the key provided at construction.  authenticate must be called before
	 * any requests can be sent to SensorCloud.
	 * @throws Exception
	 */
	public void authenticate() throws Exception
	{
		try
		{
			//determine protocol from the auth server 
			final String PROTOCOL = authServer.startsWith("http://") ? "http://" : "https://";
			
			String url = authServer+"/SensorCloud/devices/"+deviceId+"/authenticate/";
			Request request = Requests.url(url)
		                              .param("version", "1")
		                              .param("key", this.key)
		                              .accept("application/xdr")
		                              .get();
			
			// check the response code for success
			if(request.getResponseCode() != HttpURLConnection.HTTP_OK)
			{		
				throw new SensorCloudException.AuthenticationError(request.getResponseData());
			}
			
			// Extract the authentication token and server from the response
			XDRInStream xdrStream = new XDRInStream(request.getResponseStream());
			authToken = xdrStream.readString();
			apiServer = PROTOCOL + xdrStream.readString();
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("authenticate failed.", e);
		}
		
	}
	
	
	/**
	 * Determine if a sensor exists for this Device.
	 * @param sensorName
	 * @return true if the sensor exists.
	 * @throws SensorCloudException
	 */
	public boolean doesSensorExist(String sensorName) throws SensorCloudException
	{
		try
		{
			String url = apiServer + "/SensorCloud/devices/"+ deviceId + "/sensors/" + sensorName + "/";
			Request request = Requests.url(url)
	                .param("version", "1")
	                .param("auth_token", this.authToken)
	                .accept("application/xdr")
	                .get();
			
			// check the response code for success
			switch(request.getResponseCode())
			{
				case HttpURLConnection.HTTP_OK: return true;
				case HttpURLConnection.HTTP_NOT_FOUND: return false;
				default: throw new SensorCloudException("Unexpected response for get sensor. http status:" + request.getResponseCode() + " " + request.getResponseMessage() + "  info:" + request.getResponseData());
			}
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("doesSensorExist failed.", e);
		}
	}

	
	/**
	 * Add a new sensor to the Device.
	 * 
	 * @param sensorName - Name that will be given to the sensor.  Once a sensor is created it's name cannot be changed.
	 *                     sensorName must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]
	 * @throws SensorCloudException
	 */
	public void addSensor(String sensorName) throws SensorCloudException
	{
		addSensor(sensorName, "", "", "");
	}
	
	
	/**
	 * Add a new sensor to the Device.
	 * 
	 * @param sensorName - Name that will be given to the sensor.  Once a sensor is created it's name cannot be changed.
	 *                     sensorName must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]
	 * @param label - User Friendly label for the sensor.  May be changed later.
	 *                label supports unicode, and must not be more than 50 bytes when encoded as UTF-8.
	 * @param type - Type of the sensor. May be changed later.
	 *               type supports unicode, and must not be more than 50 bytes when encoded as UTF-8.
	 * @param description - Description of the sensor.  May be changed later.
	 *                      description supports unicode, and must not be more than 1000 bytes when encoded as UTF-8.
	 * 
	 * @throws SensorCloudException
	 */
	public void addSensor(String sensorName, String label, String type, String description) throws SensorCloudException
	{
		try
		{
			ByteArrayOutputStream payload = new ByteArrayOutputStream();
			XDROutStream xdrStream = new XDROutStream(payload);
			
			final int VERSION = 1;
			xdrStream.writeInt(VERSION);
			xdrStream.writeString(type);
			xdrStream.writeString(label);
			xdrStream.writeString(description);
			
			String url = apiServer + "/SensorCloud/devices/"+ deviceId + "/sensors/" + sensorName + "/";
			Request request = Requests.url(url)
	                .param("version", "1")
	                .param("auth_token", this.authToken)
	                .contentType("application/xdr")
	                .data(payload)
	                .put();
			
			// check the response code for success
			if (request.getResponseCode() != HttpURLConnection.HTTP_CREATED)
			{
				throw new SensorCloudException("AddSensorFailed. http status:" + request.getResponseCode() + " " + request.getResponseMessage() + "  info:" + request.getResponseData());
			}
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("AddSensor failed.", e);
		}
	}
	
	
	/**
	 * Determine if a channel for a specific sensor exits for the Device.
	 * 
	 * @param sensorName
	 * @param channelName
	 * 
	 * @return True if the channel exists
	 * 
	 * @throws SensorCloudException
	 */
	public boolean doesChannelExist(String sensorName, String channelName) throws SensorCloudException
	{
		try
		{
			//make a get request for channel attributes to determine if the channel exists
			String url = apiServer + "/SensorCloud/devices/"+ deviceId + "/sensors/" + sensorName + "/channels/" + channelName + "/attributes/";
			Request request = Requests.url(url)
	                .param("version", "1")
	                .param("auth_token", this.authToken)
	                .accept("application/xdr")
	                .get();
			
			// check the response code for success
			switch(request.getResponseCode())
			{
				case HttpURLConnection.HTTP_OK: return true;
				case HttpURLConnection.HTTP_NOT_FOUND: return false;
				default: throw new SensorCloudException("Unexpected response for get channel. http status:" + request.getResponseCode() + " " + request.getResponseMessage() + "  info:" + request.getResponseData());
			}
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("doesChannelExist failed.", e);
		}
	}
	
	/**
	 * Add a new channel to the Device.
	 * 
	 * @param sensorname - Name of the sensor that the channel will be added to.
	 * @param channelName - Name of the channel to create. Once a channel is created it's name cannot be changed.
	 *                      The name must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]
	 * @throws SensorCloudException
	 */
	public void addChannel(String sensorname, String channelName) throws SensorCloudException
	{
		addChannel(sensorname, channelName, "", "");
	}
	
	/**
	 * Add a new channel to the Device.
	 * 
	 * @param sensorname - Name of the sensor that the channel will be added to.
	 * @param channelName - Name of the channel to create. Once a channel is created it's name cannot be changed.
	 *                      The name must be between 1 and 50 characters in length, and only contain characters in the set[A-Z,a-z,0-9,_]
	 * @param label - User Friendly label for the channel.  May be changed later.
	 *                The label supports unicode and must not be more than 50 bytes when encoded as UTF-8.
	 * @param description - Description of the channel.  May be changed later.
	 *                      The description supports unicode, and must not be more than 1000 bytes when encoded as UTF-8.
	 * @throws SensorCloudException
	 */
	public void addChannel(String sensorname, String channelName, String label, String description) throws SensorCloudException
	{
		ByteArrayOutputStream payload = new ByteArrayOutputStream();
		XDROutStream xdrStream = new XDROutStream(payload);
		
		try
		{
			final int VERSION = 1;
			xdrStream.writeInt(VERSION);
			xdrStream.writeString(label);
			xdrStream.writeString(description);
			
			String url = apiServer + "/SensorCloud/devices/"+ deviceId + "/sensors/" + sensorname + "/channels/" + channelName + "/";
			Request request = Requests.url(url)
	                .param("version", "1")
	                .param("auth_token", this.authToken)
	                .contentType("application/xdr")
	                .data(payload)
	                .put();
			
			// check the response code for success
			if (request.getResponseCode() != HttpURLConnection.HTTP_CREATED)
			{
				throw new SensorCloudException("AddChannel failed. http status:" + request.getResponseCode() + " " + request.getResponseMessage() + "  info:" + request.getResponseData());
			}
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("AddChannel failed.", e);
		}
	}
	
	/**
	 * Add time-series data to the channel.
	 * @param sensorname - sensor containing the channel to add data to.
	 * @param channelName - channel to add data to.
	 * @param sampleRate - sample-rate for points in data
	 * @param data - a list of points to add to the channel.  The points should be ordered and match the sample-rate provided.
	 * @throws SensorCloudException
	 */
	public void addTimeseriesData(String sensorname, String channelName, SampleRate sampleRate, List<Point> data) throws SensorCloudException
	{
		try
		{
			ByteArrayOutputStream payload = new ByteArrayOutputStream();
			XDROutStream xdrStream = new XDROutStream(payload);
			
			final int VERSION = 1;
			xdrStream.writeInt(VERSION);
			xdrStream.writeInt(sampleRate.getType().getXdr());
			xdrStream.writeInt(sampleRate.getRate());
			
			//Writing an array in XDR.  an array is always prefixed by the array length
			xdrStream.writeInt(data.size());
			for(Point p: data)
			{
				xdrStream.writeHyper(p.getUnixTimestamp());
				xdrStream.writeFloat(p.getValue());
			}
			
			String url = apiServer + "/SensorCloud/devices/"+ deviceId + "/sensors/" + sensorname + "/channels/" + channelName + "/streams/timeseries/data/";
			Request request = Requests.url(url)
	                .param("version", "1")
	                .param("auth_token", this.authToken)
	                .contentType("application/xdr")
	                .data(payload)
	                .put();
		
			// check the response code for success
			if (request.getResponseCode() != HttpURLConnection.HTTP_CREATED)
			{
				throw new SensorCloudException("AddChannel failed. http status:" + request.getResponseCode() + " " + request.getResponseMessage() + "  info:" + request.getResponseData());
			}
		}
		//wrap exceptions in a SensorCloudException
		catch(IOException e)
		{
			throw new SensorCloudException("AddData failed.", e);
		}
			 
	}
}
